staging/src/k8s.io/apiserver/pkg/endpoints/filters/audit.go

```diff
@@ -160,3 +172,70 @@ func decorateResponseWriter(responseWriter http.ResponseWriter, out io.Writer, i
 	}
 	return delegate
 }
+
+var _ http.ResponseWriter = &auditResponseWriter{}
+var cloner = conversion.NewCloner()
+
+// auditResponseWriter intercepts WriteHeader, sets it in the event. If the sink is set, it will
+// create immediately an event (for long running requests).
+type auditResponseWriter struct {
+	http.ResponseWriter
+	event *auditinternal.Event
+	once  sync.Once
+	sink  chan<- *auditinternal.Event
+}
+
+func (a *auditResponseWriter) processCode(code int) {
+	a.once.Do(func() {
+		if a.sink != nil {
+			// TODO: use non-reflection-based deepcopy func when we have deepcopy for events
+			clone, err := cloner.DeepCopy(a.event)
+			if err != nil {
+				glog.Errorf("Deepcopy failed for audit event: %v", err)
```

Comment:
remove this deepcopy. That's the responsiblity of the sink now to deepcopy if it wants to keep around an object.